Implement endianness-independent serialization

ltrk2 opened the pull request:
The changes are meant to resolve the failure of functional test suite `02295_GROUP_BY_AggregateFunction` on big-endian machines by having serialization be endianness-independent.

UPDATE:
I implemented endianness-independent serialization for further aggregate functions covered by the following functional test suites:
* 00715_bounding_ratio
* 01700_deltasum
* 01762_deltasumtimestamp
* 02575_map_hashing_msan
* 02573_quantile_fuse_msan
* 02499_quantile_nan_ubsan_msan
* 02477_fuse_quantiles
* 02006_use_constants_in_with_and_select
* 00980_crash_nullable_decimal
* 00753_quantile_format
* 00700_decimal_empty_aggregates
* 00700_decimal_aggregates
* 00148_summing_merge_tree_aggregate_function
* 00181_aggregate_functions_statistics

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

ltrk2 committed:
Implement partial support for endianness-independent serialization

robot-ch-test-poll2 commented:
<!-- automatic status comment for PR #51637 from ClibMouse/ClickHouse:feature/grouparray-aggregate-function-big-endian -->
This is an automated comment for commit 2ca1b8f0a9aee279d7e7ad83beaa563c58e4c1fd with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51637/2ca1b8f0a9aee279d7e7ad83beaa563c58e4c1fd/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupArray.h:
```diff
@@ -266,19 +266,20 @@ class GroupArrayNumericImpl final
+        const size_t size = value.size();
         writeVarUInt(size, buf);
-        buf.write(reinterpret_cast<const char *>(value.data()), size * sizeof(value[0]));
+        for (const auto & element : value)
```
This is replacing a single write call with `size` write calls. While this might be necessary in big-endian machines, I expect this will make things slower for little-endian ones, which are the most common ones.

Same for reading. If you are in a little-endian arch you should be able to do the fast path of simply copying whole blocks with one operation. CH is fast because it does things in blocks :smile:

Algunenano replied:
Serialization and deserialization (both affected here) gets invoked all the time when you work with materialized views, projections or distributed queries (any query that needs to either store, retrieve or send aggregation states).

ltrk2 replied:
The reason I went with this approach is I encountered something similar in another PR (https://github.com/ClickHouse/ClickHouse/pull/50276/files#r1216984994) where @rschu1ze suggested that instead of reading the whole thing into memory and swapping items in case of big-endian, I could go one-by-one. I actually agreed with that sentiment, since in terms of maintainability going one-by-one is somewhat superior.

With that being said, I could also go with the approach of reading everything then swapping them in case of big-endian if the impact on performance was significant.

rschu1ze replied:
It depends on the kind of queries run how often serialization/deserialization is invoked. In this particular case, I would argue that aggregate function `groupArray` is rather exotic, it is not official SQL (though many DBMS seem to have it in one form or the other). Interestingly, there are a few performance tests for `groupArray` in ClickHouse, so there is a chance that performance regressions are caught.

How about this: we'll see if the performance tests complain (my bet is no), and if they do, we introduce sth. like this:

```cpp
if constexpr (std::endian::native == std::endian::little)
    buf.write(reinterpret_cast<const char *>(value.data()), size * sizeof(value[0]));
else
    for (const auto & element : value)
        writeBinaryLittleEndian(element, buf);
```

(ideally along with some comment)

Algunenano replied:
> groupArray is rather exotic

I see groupArray used quite often to replace window functions in materialized views.

> here are a few performance tests for groupArray in ClickHouse

Not sure if they are testing serialization and deserialization though.

But this is beside the point IMO. It is more about setting up a good foundation to do further changes in the future so that serialization is truly endianness independent and you can take data written by a CH and use it in another without problems. If this is done just so a few tests pass then it's the wrong reason as a) this is a breaking change (which BTW should be tagged as such for big-endianness platforms as they won't be able to read states written in the previous release) and b) I'd expect most of the serialization irks to not be covered by any tests. 

Finally, about loops vs single batches: [here is an example](https://github.com/ClickHouse/ClickHouse/pull/41089/) where changing a loop to a single read() call made uniqMerge 2x faster. These things do matter in some situations.

ltrk2 replied:
Your point around breaking change is a valid one, but the official stance is: https://github.com/ClickHouse/ClickHouse/pull/48986#issuecomment-1517038580.

The main reasons for going gradually about introducing support for interoperability is two-fold:
* multiple people are working on this so this way the effort is divided into smaller pieces,
* having to do it all at once would result in a terribly long change that'd be much harder to review.

All of this is also enabled by the fact that big-endian isn't officially supported, so backward-compatibility isn't a concern until everything comes together.

ltrk2 committed:
Implement platform-independent serialization for further aggregate functions

ltrk2 commented:
@rschu1ze I went ahead and implemented endianness-independent serialization for a few more aggregate functions that were relatively straightforward. I updated the description with the functional test suites that should cover them and I ran them locally so that they pass.

rschu1ze commented:
Restarted failing tests.
